Tests/TensorFlowTests/LazyTensorTraceTests.swift

```diff
@@ -0,0 +1,165 @@
+              %4 = Add[T: float](%2, %3)
+              %5 = Mul[T: float](%4, %3)
+            }
+            """)
+    }
+
+    func testSimpleControlFlow() {
+        let a = Tensor<Float>(5.0)
+        let addOrMul = { (useAdd: Bool, a: Tensor<Float>) in
+            useAdd ? (a + a) : (a * a)
+        }
+        let add = addOrMul(/*useAdd:*/true, a)
+        XCTAssertEqual(lazyTrace(add)!.description,
+            """
+            lazyTrace_2() -> (%1) {
+              %0 = Const[dtype: float, value: 5.0]()
+              %1 = Add[T: float](%0, %0)
+            }
+            """)
+        let mul = addOrMul(/*useAdd:*/false, a)
+        XCTAssertEqual(lazyTrace(mul)!.description,
+            """
+            lazyTrace_2() -> (%1) {
+              %0 = Const[dtype: float, value: 5.0]()
+              %1 = Mul[T: float](%0, %0)
+            }
+            """)
+    }
+
+    func testConstPromotion() {
+        let a = Tensor<Float>(10.0)
+        let b = Tensor<Float>(2.0)
+        let concreteA = a.handle.handle._tfeTensorHandle
+
+        let lazyHandle = LazyTensor(concreteA)
+        let lazyA = Tensor(handle: TensorHandle<Float>(handle: lazyHandle))
+        // Since `lazyA` is not marked as a materialized concrete
+        // tensor, this will be burnt into the trace as a constant.
+        let w1 = lazyA * b
+        let w1Trace = lazyTrace(w1)!
+        XCTAssertEqual(w1Trace.description,
+            """
+            lazyTrace_3() -> (%2) {
+              %0 = Const[dtype: float, value: 10.0]()
+              %1 = Const[dtype: float, value: 2.0]()
+              %2 = Mul[T: float](%0, %1)
+            }
+            """)
+        XCTAssertEqual(w1Trace.inputValues.count, 0)
+        let materializedHandle = LazyTensor(_materialized: concreteA)
+        let materializedLazyA = Tensor(
+            handle: TensorHandle<Float>(handle: materializedHandle))
+        // Since `materializedLazyA` is marked as a materialized concrete
+        // tensor, this will be promoted to an input for the trace.
+        let w2 = materializedLazyA * b
+        let w2Trace = lazyTrace(w2)!
+        XCTAssertEqual(w2Trace.description,
+            """
+            lazyTrace_3(%0: float) -> (%2) {
+              %1 = Const[dtype: float, value: 2.0]()
+              %2 = Mul[T: float](%0, %1)
+            }
+            """)
+        // Make sure that the promoted constants are gathered as `inputValues`.
+        XCTAssertEqual(w2Trace.inputValues.count, 1)
+        XCTAssertEqual(w2Trace.inputValues[0].valueDescription, "10.0")
+    }
+
+    private func lazyTrace<T: TensorFlowScalar>(
+        _ input: Tensor<T>) -> LazyTensorTrace? {
+        let tensor = input.handle.handle
+        guard let lazyTensor = tensor as? LazyTensor else {
+            XCTFail("Trying to get lazy trace for a non-lazy tensor.")
+            return nil
+        }
+        guard case let
+            LazyTensor.Handle.symbolic(lazyOp, _, _)  = lazyTensor.handle else {
```

Comment:
This can be on the same line as the previous line.
```suggestion
            .symbolic(lazyOp, _, _)  = lazyTensor.handle else {
```